credentials/tls/certprovider/store.go

```diff
@@ -0,0 +1,126 @@
+/*
+ *
+ * Copyright 2020 gRPC authors.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ *
+ */
+
+package certprovider
+
+import (
+	"sync"
+
+	"google.golang.org/grpc/grpclog"
+)
+
+// store is the global singleton certificate provider store.
+var store *Store
+
+func init() {
+	store = &Store{
+		providers: make(map[storeKey]*wrappedProvider),
+	}
+}
+
+// StoreKey contains data which uniquely identifies a provider instance.
+type StoreKey struct {
+	// Name is the registered name of the provider.
+	Name string
+	// Config is the configuration used by the provider instance.
+	Config StableConfig
+}
+
+// storeKey acts the key to the map of providers maintained by the store. Go
+// maps need to be indexed by comparable types, so the above StoreKey struct
+// cannot be used since it contains an interface.
+type storeKey struct {
+	// name of the certificate provider.
+	name string
+	// configuration of the certificate provider in string form.
+	config string
+}
+
+// wrappedProvider wraps a provider instance with a reference count.
+type wrappedProvider struct {
+	Provider
+	refCount int
+
+	// A reference to the key and store are also kept here to override the
+	// Close method on the provider.
+	storeKey storeKey
+	store    *Store
+}
+
+// Store is a collection of provider instances, safe for concurrent access.
+type Store struct {
+	mu        sync.Mutex
+	providers map[storeKey]*wrappedProvider
+}
+
+// GetStore returns a global singleton store of provider instances.
+func GetStore() *Store {
+	return store
+}
+
+// GetProvider returns a provider instance corresponding to key. If a provider
+// exists for key, its reference count is incremented before returning. If no
+// provider exists for key, a new is created using the registered builder. If
+// no registered builder is found, a nil provider is returned.
+func (ps *Store) GetProvider(key StoreKey) Provider {
+	ps.mu.Lock()
+	defer ps.mu.Unlock()
+
+	sk := storeKey{
+		name:   key.Name,
+		config: string(key.Config.Canonical()),
+	}
+	if wp, ok := ps.providers[sk]; ok {
+		wp.refCount++
+		return wp
+	}
+
+	b := getBuilder(key.Name)
```

Comment:
Because 
1. `getBuilder()` is static
2. type `Store` is exported, and it's fields's zero value works

I can do
```go
s := &Store{} // Instead of s := GetStore()
s.GetProvider(key)
```

Do we want to make this impossible?
Make `getBuilder()` a method of `Store`?

It might also make tests easier because we don't maintain a global state for registered builders.

Reply:
Define "works"?  It looks like there's a map lookup right before this that will panic if not initialized.

I do think we should document `Store` as: it should not be created by users; use `GetStore()` instead.

Reply:
Yes, it will panic if not initialized.
I added a comment saying users should use the `GetStore` method and not create instances of `Store` directly.

Reply:
Lookup in a nil map doesn't panic: https://play.golang.org/p/37VKa_XDCTA

Reply:
Oh yes, sorry, don't know what I was thinking.

But `GetProvider` does add the `Provider` to the map, if none was found, and if it was able to construct one successfully using the provided `Key`. So, using your own `Store` type is either useless or will panic.